src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+					RecursivelyPlanDistributedJoinNode(leftNode, query,
+													   recursivePlanningContext);
+				}
+
+				/*
+				 * Similar to LEFT JOINs, a RIGHT JOIN is recurring if the rhs
+				 * is recurring.
+				 */
+				return rightNodeRecurs;
+			}
+
+			case JOIN_FULL:
+			{
+				/*
+				 * <recurring> full join <distributed>
+				 * <distributed> full join <recurring>
+				 */
+				if (leftNodeRecurs && !rightNodeRecurs)
+				{
+					ereport(DEBUG1, (errmsg("recursively planning right side of "
+											"the full join since the other side "
+											"is a recurring rel")));
+					RecursivelyPlanDistributedJoinNode(rightNode, query,
+													   recursivePlanningContext);
+				}
+				else if (!leftNodeRecurs && rightNodeRecurs)
+				{
+					ereport(DEBUG1, (errmsg("recursively planning left side of "
+											"the full join since the other side "
+											"is a recurring rel")));
+					RecursivelyPlanDistributedJoinNode(leftNode, query,
+													   recursivePlanningContext);
+				}
+
+				/*
+				 * An OUTER JOIN is recurring if any sides of the join is
+				 * recurring. As in other outer join types, it doesn't matter
+				 * whether the other side was / became recurring or not.
+				 */
+				return leftNodeRecurs || rightNodeRecurs;
+			}
+
+			case JOIN_INNER:
+			{
+				/*
+				 * We don't need to recursively plan non-outer joins and we
+				 * already descended into sub join trees to handle outer joins
+				 * buried in them.
+				 */
+				return leftNodeRecurs && rightNodeRecurs;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("got unexpected join type (%d) when recursively "
+									   "planning a join",
+									   joinExpr->jointype)));
+			}
+		}
+	}
+	else
+	{
+		ereport(ERROR, errmsg("got unexpected node type (%d) when recursively "
+							  "planning a join",
+							  nodeTag(node)));
+	}
+}
+
+
+/*
+ * RecursivelyPlanDistributedJoinNode is a helper function for
+ * RecursivelyPlanRecurringTupleOuterJoinWalker that recursively plans given
+ * distributed node that is known to be inner side of an outer join.
+ *
+ * We call a node "distributed" if it points to a distributed table or a
+ * more complex object (i.e., a join tree or a subquery) that can be pushed
+ * down to the worker nodes directly. For a join, this means that it's either
+ * an INNER join where any side of it is a distributed table / a distributed
+ * sub join tree, or an OUTER join where the outer side is a distributed table
+ * / a distributed sub join tree.
+ */
+static void
+RecursivelyPlanDistributedJoinNode(Node *node, Query *query,
+								   RecursivePlanningContext *recursivePlanningContext)
+{
+	if (IsA(node, JoinExpr))
+	{
+		/*
+		 * This, for example, means that RecursivelyPlanRecurringTupleOuterJoinWalker
+		 * needs to plan inner side, i.e., "<distributed> INNER JOIN <distributed>",
+		 * of the following join:
+		 *   <recurring> LEFT JOIN (<distributed> INNER JOIN <distributed>)
```

Comment:
can't we have `<recurring> LEFT JOIN (<distributed> **LEFT*** JOIN <distributed>` here?

Do we have any such test? If not, would be good to add

Reply:
> can't we have <recurring> LEFT JOIN (<distributed> **LEFT*** JOIN <distributed> here?

yes we can have, let me add such a test too